wireguard-go-rs/src/lib.rs

```diff
@@ -1,70 +1,259 @@
+            let config_bytes = unsafe { slice::from_raw_parts_mut(ptr, config_len) };
+            config_bytes.zeroize();
+        }
+
+        // SAFETY: the pointer was created by wgGetConfig, and we are no longer using it.
+        unsafe { ffi::wgFreePtr(ptr.cast()) };
+
+        Some(t)
+    }
+
+    pub fn set_config(&self, config: &CStr) -> Result<(), Error> {
+        // SAFETY: pointer is valid for the lifetime of this function.
+        let code = unsafe { ffi::wgSetConfig(self.handle, config.as_ptr()) };
+        result_from_code(code)
+    }
+
     #[cfg(daita)]
-    pub fn wgActivateDaita(
-        tunnel_handle: i32,
-        peer_public_key: *const u8,
-        machines: *const c_char,
+    pub fn activate_daita(
+        &self,
+        peer_public_key: &[u8; 32],
+        machines: &CStr,
         events_capacity: u32,
         actions_capacity: u32,
-    ) -> i32;
+    ) -> Result<(), Error> {
+        // SAFETY: pointers are valid for the lifetime of this function.
+        let code = unsafe {
+            ffi::wgActivateDaita(
+                self.handle,
+                peer_public_key.as_ptr(),
+                machines.as_ptr(),
+                events_capacity,
+                actions_capacity,
+            )
+        };
 
-    // Frees a pointer allocated by the go runtime - useful to free return value of wgGetConfig
-    pub fn wgFreePtr(ptr: *mut c_void);
+        result_from_code(code)
+    }
 
-    // Returns the file descriptor of the tunnel IPv4 socket.
+    /// Get the file descriptor of the tunnel IPv4 socket.
     #[cfg(target_os = "android")]
-    pub fn wgGetSocketV4(handle: i32) -> Fd;
+    pub fn get_socket_v4(&self) -> Fd {
+        unsafe { ffi::wgGetSocketV4(self.handle) }
+    }
 
-    // Returns the file descriptor of the tunnel IPv6 socket.
+    /// Get the file descriptor of the tunnel IPv6 socket.
     #[cfg(target_os = "android")]
-    pub fn wgGetSocketV6(handle: i32) -> Fd;
+    pub fn get_socket_v6(&self) -> Fd {
+        unsafe { ffi::wgGetSocketV6(self.handle) }
+    }
+}
+
+impl Drop for Tunnel {
+    fn drop(&mut self) {
+        let code = unsafe { ffi::wgTurnOff(self.handle) };
+        if let Err(e) = result_from_code(code) {
+            log::error!("Failed to stop wireguard-go tunnel,oerror_code={code} ({e:?})")
+        }
+    }
+}
+
+fn result_from_code(code: i32) -> Result<(), Error> {
+    // NOTE: must be kept in sync with enum definition
+    Err(match code {
+        0.. => return Ok(()),
+        -1 => Error::GeneralFailure,
+        -2 => Error::IntermittentFailure,
+        -3 => Error::UnknownTunnel,
+        -4 => Error::UnknownPeer,
+        -5 => Error::EnableDaita,
+        _ => Error::Other,
+    })
+}
+
+impl Error {
+    pub const fn as_raw(self) -> i32 {
+        self as i32
+    }
+}
+
+mod ffi {
+    use super::{Fd, LoggingCallback};
+    use core::ffi::{c_char, c_void};
+
+    extern "C" {
```

Comment:
Should the error cases for these function be documented here?

Reply:
Maybe, but that seems like a lot of work and hard to keep in sync

Reply:
Up to you. I'm ok with just refering to the libwg.go source code for details also.